File: references/openguardrails-main/openguardrails-main/integrations/agent/langgraph/tests/mock_runtime.py
"""The strict mock runtime — the v0.8 contract made executable, offline.

It validates every received GuardEvent against the schema's EXACT required
set (``additionalProperties: false`` — nine required fields plus the one
optional ``integration``,
nothing extra) and RECORDS violations instead of merely 400-ing, because a
fail-open integration would swallow a 400 and a wire regression would
otherwise pass silently. Tests assert ``runtime.violations == []``.
"""

from __future__ import annotations

import json
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

API_KEY = "ogr_test_key"

# The v0.8 GuardEvent: schema/guard-event.schema.json `required`, verbatim.
REQUIRED_FIELDS = {
    "kind",
    "step_id",
    "agent_id",
    "agent_type",
    "agent_workspace",
    "agent_user",
    "llm_protocol",
    "payload",
}

#: The one OPTIONAL field (2026-08-17): ``integration``, the reporter's own
#: ``"name/version"``. An ALLOWLIST, not a relaxation — an unknown key is still a
#: violation; only a MISSING ``integration`` stopped being one, which is what lets
#: a runtime and a reporter roll forward independently.
OPTIONAL_FIELDS = {"integration"}

KINDS = {"step/request", "step/response"}
LLM_PROTOCOLS = {"openai.chat", "openai.responses", "anthropic.messages", "canonical"}
FIVE_TUPLE = ("agent_id", "agent_type", "agent_workspace", "agent_user")


class MockRuntime:
    """A scriptable /v1/evaluate + /v1/heartbeat server on a loopback port."""

    def __init__(self) -> None:
        self.events: list[dict] = []
        self.heartbeats: list[dict] = []
        self.violations: list[str] = []
        self.verdicts: list[dict] = []  # FIFO script; empty → default allow
        self.fail_statuses: list[int] = []  # FIFO of statuses to answer with first
        self._served = 0

        runtime = self

        class Handler(BaseHTTPRequestHandler):
            def log_message(self, *args) -> None:  # keep pytest output clean
                pass

            def _reply(self, status: int, body: dict) -> None:
                data = json.dumps(body).encode("utf-8")
                self.send_response(status)
                self.send_header("Content-Type", "application/json")
                self.send_header("Content-Length", str(len(data)))
                self.end_headers()
                self.wfile.write(data)

            def do_POST(self) -> None:
                body = json.loads(self.rfile.read(int(self.headers["Content-Length"] or 0)))
                if self.headers.get("Authorization") != f"Bearer {API_KEY}":
                    runtime.violations.append(f"bad auth: {self.headers.get('Authorization')!r}")
                    self._reply(401, {"error": "unauthorized"})
                    return
                if runtime.fail_statuses:
                    self._reply(runtime.fail_statuses.pop(0), {"error": "scripted_failure"})
                    return
                if self.path == "/v1/heartbeat":
                    runtime.heartbeats.append(body)
                    self._reply(200, {"ok": True})
                    return
                if self.path != "/v1/evaluate":
                    self._reply(404, {"error": "not_found"})
                    return
                runtime._validate(body)
                runtime.events.append(body)
                runtime._served += 1
                verdict = runtime.verdicts.pop(0) if runtime.verdicts else {"decision": "allow"}
                verdict.setdefault("event_id", f"evt-{runtime._served}")
                verdict.setdefault("provider", "mock-runtime")
                self._reply(200, verdict)

        self._server = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
        self.url = f"http://127.0.0.1:{self._server.server_address[1]}"
        self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._thread.start()

    def _validate(self, event: dict) -> None:
        keys = set(event)
        if keys < REQUIRED_FIELDS or keys - REQUIRED_FIELDS - OPTIONAL_FIELDS:
            self.violations.append(
                f"field set mismatch: missing={sorted(REQUIRED_FIELDS - keys)} "
                f"extra={sorted(keys - REQUIRED_FIELDS - OPTIONAL_FIELDS)}"
            )
            return
        if event["kind"] not in KINDS:
            self.violations.append(f"bad kind: {event['kind']!r}")
        if not isinstance(event["step_id"], str) or not event["step_id"]:
            self.violations.append(f"bad step_id: {event['step_id']!r}")
        for field in FIVE_TUPLE:
            if not isinstance(event[field], str):
                self.violations.append(f"{field} is not a string: {event[field]!r}")
        if event["llm_protocol"] not in LLM_PROTOCOLS:
            self.violations.append(f"bad llm_protocol: {event['llm_protocol']!r}")
        if not isinstance(event["payload"], dict):
            self.violations.append(f"payload is not an object: {type(event['payload'])}")

    def close(self) -> None:
        self._server.shutdown()
        self._server.server_close()

File: references/openguardrails-main/openguardrails-main/integrations/agent/codex/test/mock-runtime.mjs
// A STRICT v0.8 mock OGR runtime (node:http, offline), shared by both hook
// suites. It rejects any GuardEvent deviating from
// schema/guard-event.schema.json — the nine required fields plus the one
// optional `integration`, nothing else, no retired v0.6/v0.7 fields — and records every violation so a wire
// regression fails the run loudly instead of hiding behind fail-open.
import { createServer } from "node:http"

export const API_KEY = "ogr_test"

// The required set from schema/guard-event.schema.json. The schema also has
// additionalProperties:false, so the only keys allowed beyond these are the
// optional ones below.
const EVENT_KEYS = [
  "kind", "step_id", "agent_id", "agent_type", "agent_workspace",
  "agent_user", "llm_protocol", "payload",
].sort()

// The one OPTIONAL field (2026-08-17): `integration`, the reporter's own
// "name/version". An ALLOWLIST, not a relaxation — an unknown key is still a
// violation; only a MISSING `integration` stopped being one, which is what lets
// a runtime and a reporter roll forward independently.
const OPTIONAL_EVENT_KEYS = ["integration"]

function validateEvent(ev) {
  const errs = []
  const keys = Object.keys(ev).sort()
  const missing = EVENT_KEYS.filter((k) => !keys.includes(k))
  const extra = keys.filter((k) => !EVENT_KEYS.includes(k) && !OPTIONAL_EVENT_KEYS.includes(k))
  if (missing.length || extra.length) {
    errs.push(`key set is [${keys}], missing [${missing}], unexpected [${extra}]`)
  }
  if (!["step/request", "step/response"].includes(ev.kind)) errs.push(`kind ${ev.kind}`)
  if (typeof ev.step_id !== "string" || !ev.step_id) errs.push("step_id must be a non-empty string")
  for (const f of ["agent_id", "agent_type", "agent_workspace", "agent_user"]) {
    if (typeof ev[f] !== "string") errs.push(`${f} must be a string`)
  }
  if (!["openai.chat", "openai.responses", "anthropic.messages", "canonical"].includes(ev.llm_protocol)) {
    errs.push(`llm_protocol ${ev.llm_protocol}`)
  }
  // Presence is optional on the WIRE; for THESE hooks it is mandatory — the mock
  // tolerating a missing `integration` must not let ours stop sending it.
  if (!/^ogr-codex(-automode)?\/[0-9]/.test(String(ev.integration))) {
    errs.push(`integration ${ev.integration}`)
  }
  if (typeof ev.payload !== "object" || ev.payload === null || Array.isArray(ev.payload)) {
    errs.push("payload must be an object")
  }
  return errs
}

export async function startMockRuntime() {
  const state = {
    verdictHandler: () => ({
      status: 200,
      body: { event_id: "evt_1", provider: "mock-runtime", decision: "allow" },
    }),
    requests: [],
    violations: [],
  }
  const server = createServer((req, res) => {
    let raw = ""
    req.on("data", (c) => (raw += c))
    req.on("end", () => {
      const body = raw ? JSON.parse(raw) : {}
      state.requests.push({ path: req.url, body, auth: req.headers.authorization ?? "" })
      if (req.url !== "/v1/evaluate") {
        state.violations.push(`unexpected path ${req.url} (v0.8 has no other event path)`)
        res.writeHead(404).end()
        return
      }
      const errs = validateEvent(body)
      if (req.headers.authorization !== `Bearer ${API_KEY}`) errs.push(`bad auth '${req.headers.authorization}'`)
      if (errs.length) {
        state.violations.push(...errs)
        res.writeHead(400, { "content-type": "application/json" })
        res.end(JSON.stringify({ error: "invalid_event", details: errs }))
        return
      }
      const { status, body: out } = state.verdictHandler(body)
      res.writeHead(status, { "content-type": "application/json" })
      res.end(JSON.stringify(out))
    })
  })
  await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve))
  state.url = `http://127.0.0.1:${server.address().port}`
  state.close = () => server.close()
  return state
}

export const allowVerdict = (extra = {}) => () => ({
  status: 200,
  body: { event_id: "evt_1", provider: "mock-runtime", decision: "allow", ...extra },
})
